zcash_client_sqlite/src/wallet/commitment_tree.rs

```diff
@@ -257,23 +262,29 @@ type Error = Either<io::Error, rusqlite::Error>;
 pub(crate) fn get_shard<H: HashSer>(
     conn: &rusqlite::Connection,
     table_prefix: &'static str,
-    shard_root: Address,
+    shard_root_addr: Address,
 ) -> Result<Option<LocatedPrunableTree<H>>, Error> {
     conn.query_row(
         &format!(
-            "SELECT shard_data
+            "SELECT shard_data, root_hash
              FROM {}_tree_shards
              WHERE shard_index = :shard_index",
             table_prefix
         ),
-        named_params![":shard_index": shard_root.index()],
-        |row| row.get::<_, Vec<u8>>(0),
+        named_params![":shard_index": shard_root_addr.index()],
+        |row| Ok((row.get::<_, Vec<u8>>(0)?, row.get::<_, Option<Vec<u8>>>(1)?)),
     )
     .optional()
     .map_err(Either::Right)?
-    .map(|shard_data| {
+    .map(|(shard_data, root_hash)| {
         let shard_tree = read_shard(&mut Cursor::new(shard_data)).map_err(Either::Left)?;
-        Ok(LocatedPrunableTree::from_parts(shard_root, shard_tree))
+        let located_tree = LocatedPrunableTree::from_parts(shard_root_addr, shard_tree);
+        if let Some(root_hash_data) = root_hash {
+            let root_hash = H::read(Cursor::new(root_hash_data)).map_err(Either::Left)?;
+            Ok(located_tree.reannotate_root(Some(Rc::new(root_hash))))
```

Comment:
What does this do? #831 gave me the impression that the `root_hash` column was only being stored for caching purposes, and that it was re-derivable from `shard_data`.

Reply:
The root can only be re-derived from `shard_data` if the subtree is complete; if there are any `nil` children then we can't re-derive the root, so we have to store it as an annotation.